### PR TITLE
Allow plugins to look up tools from the toolchain (such as `docc`)

### PR DESCRIPTION
**Pull request by abertelrud**
Add ability to pass down a list of additional search paths to a plugin, and use them when looking up the plugin.  Configure call sites to pass down the toolchain bin directory.

### Motivation:

This implements part of the design of plugins: they have access to binary executables and built executables that they declare dependencies on, plus what is in the toolchain.

### Modifications:

- Add ability to pass down a list of additional search paths to a plugin, and use them when looking up the plugin.
- Configure call sites to pass down the toolchain bin directory.
- Extend a unit test to check that it works.

### CODE REVIEW

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]

File: Sources/PackagePlugin/Context.swift [hunk 1]
```diff
@@ -39,14 +41,32 @@ public struct PluginContext {
...
+            return Tool(name: name, path: path)
+        } else {
+            for dir in toolSearchDirectories {
+#if os(Windows)
+                let hostExecutableSuffix = ".exe"
+#else
+                let hostExecutableSuffix = ""
+#endif
+                let path = dir.appending(name + hostExecutableSuffix)
+                if FileManager.default.isExecutableFile(atPath: path.string) {
```
Comment (tomerd) [inline Sources/PackagePlugin/Context.swift:54]: should this be an addition / exertion on FileSystem instead hard coded to local one?

File: Tests/FunctionalTests/PluginTests.swift [hunk 2]
```diff
@@ -283,9 +288,11 @@ class PluginTests: XCTestCase {
                 func pluginEmittedOutput(_ data: Data) {
                     dispatchPrecondition(condition: .onQueue(delegateQueue))
                     outputData.append(contentsOf: data)
+                    print("🧩 \(String(decoding: data, as: UTF8.self))")
```
Comment (tomerd) [inline Tests/FunctionalTests/PluginTests.swift:291]: fancy 😍

File: Sources/Commands/SwiftTool.swift [hunk 3]
```diff
@@ -758,6 +758,9 @@ public class SwiftTool {
             // binary targets.
             let dataDir = try self.getActiveWorkspace().location.workingDirectory
             let builtToolsDir = dataDir.appending(components: try self._hostToolchain.get().triple.platformBuildPathComponent(), buildEnvironment.configuration.dirname)
+            
+            // Use the directory containing the compiler as an additional search directory.
+            let toolSearchDirs = [try self.getToolchain().swiftCompilerPath.parentDirectory]
```
Comment (tomerd) [inline Sources/Commands/SwiftTool.swift:763]: should this made into a func so that can be reused in SwiftPackageTool or other tools that may need this?

**Pull request merged**